Update gen_pipeline to accept team target name and secrets file

jmcatamney opened the pull request:
Update gen_pipeline to take a team target for the pipeline_type parameter, which then uses it to automatically construct the appropriate fly command.

This enables us to simply use pipeline_type which is the -t argument (e.g. prod, dev, cm) as the fly target type. This helps clean up the code slightly, and keep the target short since we already know we are in GPDB.

Notes:
- This removes the “gpdb-” prefix currently in use for the target name.  This may require updating your flyrc file.
- We've extracted the branch name into a separate variable to make this easier to backport and to modify in future when we branch new versions.
- The fly command is now a single line, as we don't expect the command to need to be edited once it's generated.
- See #7720 for previous discussion on this topic.

jchampio commented:
> The fly command is now a single line, as we don't expect the command to need to be edited once it's generated.

FWIW, I think the newlines make it easier to do a quick sanity check as I'm copy-pasting.

jchampio commented on concourse/pipelines/gen_pipeline.py:
```diff
@@ -78,11 +83,12 @@
     default_remote = "<https://github.com/<github-user>/gpdb>"
 
-    remote = subprocess.check_output("git ls-remote --get-url", shell=True).rstrip()
+    remote = subprocess.check_output("git ls-remote --get-url".split()).rstrip()
```
I tend to prefer explicit lists to the `"cmd arg arg".split()` construction, if only because `split()` will eventually go terribly wrong if the command evolves:
```python
subprocess.check_output("cmd 'quoted arg'".split()) # [ "cmd", "'quoted", "arg'" ]
```

jchampio commented on concourse/pipelines/gen_pipeline.py:
```diff
@@ -94,21 +100,27 @@ def suggested_git_remote():
+    parts = branch.split("X_STABLE", 2)
+    # if first part is a one or two digit number
+    return len(parts[0]) < 3 and parts[0].isdigit() \
+           and parts[1] is not None  # if second part ends in X_STABLE
```
I think you'll hit an `IndexError` in some cases; `split()` won't give you `None` in its list. E.g. try `10` as a branch name.

kalensk replied:
Pushed a fixup that uses a best effort regex.

marksliva commented:
nit: the newly supported cm target generated files like `gpdb-cm-pivotal.yml` should be added to the `concourse/pipelines/.gitignore` file. Or we can generalize the existing exclude to also match it.

jchampio commented on concourse/pipelines/gen_pipeline.py:
```diff
@@ -142,6 +154,15 @@ def validate_pipeline_release_jobs(raw_pipeline_yml):
+
+def validate_target(target):
+    secrets_filenames = [os.path.splitext(os.path.basename(secrets_file))[0] for secrets_file in glob.glob("%s/ccp_ci_secrets_*.yml" % SECRETS_PATH)]
+    targets_in_secrets_filenames = [secrets_file.split("ccp_ci_secrets_")[1] for secrets_file in secrets_filenames]
```
These are very long lines.

bradfordboyle replied:
Wouldn't it be simpler to form the expected path (e.g., "<secrets-path>/ccp_ci_secrets_<target>.yml") and check if the file exists?

marksliva commented:
LGTM once Jacob's comments are addressed.

bradfordboyle commented on concourse/pipelines/gen_pipeline.py:
```diff
@@ -94,21 +100,27 @@ def suggested_git_remote():
-    if branch == "master" or branch == "5X_STABLE":
-        return default_branch
-    return branch
+def is_a_released_branch(branch):
```
I am not sure this logic is correct. Do we consider `5.10.X` a released branch? Or just `5X_STABLE` and `6X_STABLE`? If the later, why not just `branch.endswith('X_STABLE')`?

kalensk replied:
Pushed a fixup that uses a best effort regex.

I am not sure what the valid release branches are or will be. I will leave that to releng to update the function as needed. This was simply an effort to get out of hardcoding every release branch and remembering to update every reference to "master". Further improvements after this PR is totally welcomed.

Using `branch.endswith('X_STABLE')` is a bit tricky since it would not work for the following branch name, `165187614_my_feature_6X_STABLE`.

This function should return false for the following cases: `6`, `123_my_feature_6X_STABLE`, possibly `60X_STABLE`, etc. 
And this function should return true for the following cases: `5X_STABLE`, `6X_STABLE`, etc.

bradfordboyle commented on concourse/pipelines/gen_pipeline.py:
```diff
@@ -155,12 +176,12 @@ def create_pipeline(args):
 
     pipeline_yml = render_template(args.template_filename, context)
-    if args.pipeline_type == 'prod':
+    if args.pipeline_target == 'prod':
```
Not the fault of this PR, but there are some jobs in the pipeline that only get included if we are flying a production pipeline. What about the case where I want a dev version of the production pipeline on a team Concourse instance?

Would it make more sense to have a separate arg for the target? This would allow

```sh
./gen_pipeline.py --type=prod --target=releng # dev version of the prod pipeline on a team concourse instance
./gen_pipeline.py --type=prod # prod version of the prod pipeline on the prod concourse instance
```

kalensk replied:
I was unaware of that use case. How often is it used, and is it "really" needed? Obviously someone can copy and edit the generated pipeline files by hand achieve the above goal (which is not user friendly).

One of the initial goals of normalizing the secret files and removing the "gpdb-" prefix was to avoid having yet another argument.

bradfordboyle commented on concourse/pipelines/gen_pipeline.py:
```diff
@@ -173,50 +194,71 @@ def create_pipeline(args):
+        secrets += "-l %s/%s " % (SECRETS_PATH, secret)
+
+    return '''fly -t %s \
+set-pipeline \
```
I feel like this is a case where using the `format()` method of a string would be helpful

```py
args = {
    'target': <target>,
    'name': <name>,
   ...
}
'fly -t {target} -p {name} ...'.format(**args)
```

bradfordboyle commented on concourse/pipelines/gen_pipeline.py:
```diff
@@ -46,6 +47,10 @@
     extensions=['jinja2.ext.loopcontrols']
 )
 
+BRANCH = "master"  # when branching gpdb update to 7X_STABLE, 6X_STABLE, etc.
```
Nitpick: Would naming this `BASE_BRANCH` sense? My first thought was that we were hard-coding the branch name.

kalensk commented:
> FWIW, I think the newlines make it easier to do a quick sanity check as I'm copy-pasting.

I prefer the single line as it enables me to quickly double click and then easily paste into my terminal, whereas the multi-line was a bit more cumbersome to copy and use. I might argue that if one needs to continual do sanity checks (albeit quick), perhaps that is a code smell for improving the tool?

goutamtadi1 commented on concourse/pipelines/gen_pipeline.py:
```diff
@@ -94,21 +100,26 @@ def suggested_git_remote():
-    if branch == "master" or branch == "5X_STABLE":
-        return default_branch
-    return branch
+def is_a_released_branch(branch):
```
At first I thought `is_a_released_branch` refers some branch like `5.20.0` branch on Greenplum. I guess this should be named as `is_a_release_branch`. I know that's a simple naming update, however I felt this is much clearer.

jmcatamney commented:
@bradfordboyle and @jchampio: We've addressed PR comments and pushed them, let us know if everything looks good.

In particular, as per Bradford's suggestion, we added a new flag to specify pipeline configuration independent of target, squashed into the "rename pipeline_target" commit.  Now, adding `-c prod` makes a dev or team pipeline that looks like the prod pipeline, `-c full` enables all platforms and test sections but leaves out prod-specific release stuff, and `-c minimal` does the current ICW-on-centos6 default.  How's that sound?

jchampio commented on concourse/pipelines/gen_pipeline.py:
```diff
@@ -286,14 +286,24 @@ def main():
+        dest='pipeline_configuration',
+        default='minimal',
+        help='Set of platforms and test sections to use; ignored if target is prod.  Valid options are prod (same as the prod pipeline), '
+             'full (everything except release jobs), and minimal (centos6 and ICW).'
```
nit: it looks like `minimal` doesn't really do what this documentation says; it just means "use whatever you passed in `-A` and `-O`." Would something like `none` make more sense?

kalensk replied:
Also, I think phrase `ignored if target is prod` is outdated since it does work. Also, it sounded like releng would like to generate a prod pipeline on a "dev" concourse target for testing.

slari-pivotal replied:
@jchampio We'll change it to "default."

@kalensk You can use `-t dev -c prod` to create a prod-like pipeline on a dev instance, what you can't do is use `-t prod -c default` to change the actual master pipeline.  We can update the wording to make that more clear.

kalensk commented on concourse/pipelines/gen_pipeline.py:
```diff
@@ -235,7 +276,7 @@ def main():
     )
 
-    default_output_filename = "gpdb_master-generated.yml"
+    default_output_filename = "gpdb_%s-generated.yml" % BASE_BRANCH
```
total nitpick: perhaps adding default_output_filename near the top of main, or adding a space between these lines to break it up visually. (It was a bit hard to quickly find where default_output_filename was defined since it was blending into the parser.add_argument line...)

kalensk commented:
Nit: Can we update the help messages for the -o, -u, and -b flags to be more clear? That is each should be more clear in that they update _both_ the pipeline name and pipeline filename, with their respective options.

kalensk commented:
Good job, love the "full" option. Thanks!
